internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	token, err := conf.Token(context.WithValue(ctx, oauth2.HTTPClient, httpClient))
+	if err != nil {
+		return nil, fmt.Errorf("error fetching initial oidc client_credentials token: %w", err)
+	}
+
+	opts.Logger.Info("Fetched initial oidc client_credentials token")
+
+	return &ClientCredentials{
+		log:          opts.Logger,
+		currentToken: token,
+		httpClient:   httpClient,
+		fetchTokenFn: conf.Token,
+	}, nil
+}
+
+func toConfig(opts ClientCredentialsOptions) (*ccreds.Config, *http.Client, error) {
+	scopes := opts.Scopes
+	if len(scopes) == 0 {
+		// If no scopes are provided, then the default is to use the 'openid' scope
+		// since that is always required for OIDC so implicitly add it.
+		scopes = []string{oidcScopeOpenID}
+	}
+
+	var oidcScopeFound bool
+	for _, scope := range scopes {
+		if scope == oidcScopeOpenID {
+			oidcScopeFound = true
+			break
+		}
+	}
+	if !oidcScopeFound {
+		return nil, nil, fmt.Errorf("oidc client_credentials token source requires the %q scope", oidcScopeOpenID)
+	}
+
+	tokenURL, err := url.Parse(opts.TokenURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing token URL: %w", err)
+	}
+	if tokenURL.Scheme != "https" {
+		return nil, nil, fmt.Errorf("OIDC token provider URL requires 'https' scheme: %q", tokenURL)
+	}
+
+	conf := &ccreds.Config{
+		ClientID:     opts.ClientID,
+		ClientSecret: opts.ClientSecret,
+		TokenURL:     opts.TokenURL,
+		Scopes:       scopes,
+	}
+
+	if len(opts.Audiences) == 0 {
+		return nil, nil, errors.New("oidc client_credentials token source requires at least one audience")
+	}
+
+	conf.EndpointParams = url.Values{"audience": opts.Audiences}
+
+	// If caPool is nil, then the Go TLS library will use the system's root CA.
+	var caPool *x509.CertPool
+	if len(opts.CAPEM) > 0 {
+		caPool = x509.NewCertPool()
+		if !caPool.AppendCertsFromPEM(opts.CAPEM) {
+			return nil, nil, errors.New("failed to parse CA PEM")
+		}
+	}
+
+	return conf, &http.Client{
+		Timeout: time.Second * 30,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				MinVersion: tls.VersionTLS12,
+				RootCAs:    caPool,
+			},
+		},
+	}, nil
+}
+
+func (c *ClientCredentials) Token() (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	if !c.currentToken.Valid() {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		if err := c.renewToken(ctx); err != nil {
+			return "", err
+		}
+	}
+
+	return c.currentToken.AccessToken, nil
+}
+
+func (c *ClientCredentials) renewToken(ctx context.Context) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// We need to check if the current token is valid because we might have lost
+	// the mutex lock race from the caller and we don't want to double-fetch a
+	// token unnecessarily!
+	if c.currentToken.Valid() {
+		return nil
+	}
+
+	token, err := c.fetchTokenFn(context.WithValue(ctx, oauth2.HTTPClient, c.httpClient))
+	if err != nil {
+		return err
+	}
+
+	c.currentToken = token
```

Comment:
Should check that `token.Valid` before saving it? As a sanity check

Reply:
:+1:  Can't hurt :slightly_smiling_face: